oeps/oep-0050-hooks-extension-framework.rst

```diff
@@ -0,0 +1,288 @@
+https://github.com/edx/edx-platform/tree/master/docs/guides. This location
+should contain a document that covers the design principles and details that did
+not land in the OEP as well as the final decisions made in the OEP. Also will
+index the list of hooks and their current status. It should not cover the
+history and all the previous states of the framework. It should reflect the
+latest way of thinking and examples of how to use the framework.
+
+Given the public promise made by the framework, the signature of the functions
+(input/output) of every hook should be maintained in a stable way. The same
+applies for additions or removals from the list of public triggers, which should
+follow a path of deprecation first and preferably be given an optional hook
+where the same context information or code execution path can be accessed.
+
+Registering more than one function for a hook should be possible. This creates
+two different interfaces given that filters are expected to return something and
+the platform should act accordingly. Due to this different nature, the mechanism
+by which a developer connects to a hook differs from events to filters.
+
+Events will be called via Django signals. There will be a new library that
+defines a list of semantically meaningful signals based on higher-level events
+and platform concepts. This library will be imported at the platform code and
+will, when possible, replace the current signals the platform already emits.
+Developers that want to use a hook, should also import this library at the
+plugins where they are writing the functions that connect to the event.
+
+For filters, the possibility of registering more than one function makes some
+sort of reducing routine for the results necessary. Therefore the proposal is to
+make them work as a pipeline, so the results of the previously executed function
+should be made available to the next function. Doing so makes the calling order
+of its registered functions important and thus it must be respected. This is not
+supported by Django signals and for this case, the mechanism to register
+functions will be the Django settings module.
+
+When possible, hooks should be compatible with the Async Event Messaging
+described in OEP-41.
+In terms of error handling, the implementation should allow the configuration of
+silent failures letting operators know the source of the error.
+
+
+Rationale
+=========
+
+Hooks are a way for one piece of code to interact/modify another piece of code
+at specific, pre-defined spots. Many of the business rules of an installation
+can be altered and extended if the right spots are allowed to be extended by
+open edX plugins. Initially, this proposal is meant as a way for community
+members to extend the platform, but it could even become a way for the
+installation at edx.org to implement some of its particular business rules in a
+way that is sufficiently separated from the Open edX core.
```

Comment:
[+1] Yes, completely agree. Once this framework is in place, it would be good to follow-up with technical refactorings of edx.org specific business logic.